### PR TITLE
[Merged by Bors] - change the order of operation in zsmulRec and nsmulRec

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
We change the following field in the definition of an additive commutative monoid:
```diff
 nsmul_succ : ∀ (n : ℕ) (x : G),
-  AddMonoid.nsmul (n + 1) x = x + AddMonoid.nsmul n x
+  AddMonoid.nsmul (n + 1) x = AddMonoid.nsmul n x + x
```
where the latter is more natural

We adjust the definitions of `^` in monoids, groups, etc.
Originally there was a warning comment about why this natural order was preferred

> use `x * npowRec n x` and not `npowRec n x * x` in the definition to make sure that
> definitional unfolding of `npowRec` is blocked, to avoid deep recursion issues.

but it seems to no longer apply.

Remarks on the PR : 
* `pow_succ` and `pow_succ'` have switched their meanings.
* Most of the time, the proofs were adjusted by priming/unpriming one lemma, or exchanging left and right; a few proofs were more complicated to adjust.
* In particular, [Mathlib/NumberTheory/RamificationInertia.lean] used `Ideal.IsPrime.mul_mem_pow` which is defined in [Mathlib/RingTheory/DedekindDomain/Ideal.lean]. Changing the order of operation forced me to add the symmetric lemma `Ideal.IsPrime.mem_pow_mul`. 
* the docstring for Cauchy condensation test in [Mathlib/Analysis/PSeries.lean] was
mathematically incorrect, I added the mention that the function is antitone.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Can you fix the description of the PR? Currently it seems to say a false statement.

**Comment by AntoineChambert-Loir** [comment]
I updated the description! Sorry for the typo.

**Comment by AntoineChambert-Loir** [comment]
Merging is tricky because it seems that something disappeared from Mathlib/Algebra/GroupPower/Order !

File: Mathlib/Algebra/GradedMonoid.lean [hunk 1]
```diff
@@ -207,8 +207,9 @@ theorem gnpowRec_zero (a : GradedMonoid A) : GradedMonoid.mk _ (gnpowRec 0 a.snd
 
 @[simp]
 theorem gnpowRec_succ (n : ℕ) (a : GradedMonoid A) :
-    (GradedMonoid.mk _ <| gnpowRec n.succ a.snd) = a * ⟨_, gnpowRec n a.snd⟩ :=
+    (GradedMonoid.mk _ <| gnpowRec n.succ a.snd) = ⟨_, gnpowRec n a.snd⟩ * a :=
   Sigma.ext (succ_nsmul _ _) (heq_of_cast_eq _ rfl).symm
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/GradedMonoid.lean:212]:
```suggestion
```

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 2]
```diff
@@ -264,6 +264,11 @@ instance pow [One α] [Pow α ℕ] : Pow (WithZero α) ℕ :=
     | none, _ + 1 => 0
     | some x, n => ↑(x ^ n)⟩
 
+example [Monoid α] (n : ℕ) :
+    (0 : WithZero α) ^ (n + 1) = (0 : WithZero α) ^ n * 0 :=  by
+  simp only [mul_zero]; rfl
+
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:271]: I assume this can be deleted

File: Mathlib/Algebra/Group/WithOne/Defs.lean [hunk 3]
```diff
@@ -279,7 +284,9 @@ instance monoidWithZero [Monoid α] : MonoidWithZero (WithZero α) :=
       | some x => congr_arg some <| pow_zero x,
     npow_succ := fun n x =>
       match x with
-      | none => rfl
+      | none => by
+        change 0 ^ (n + 1) = 0 ^ n * 0
+        simp only [mul_zero]; rfl
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/WithOne/Defs.lean:289]: I'm slightly surprised the defeq broke here

File: Mathlib/Algebra/TrivSqZeroExt.lean [hunk 4]
```diff
@@ -658,6 +658,11 @@ theorem inl_pow [Monoid R] [AddMonoid M] [DistribMulAction R M] [DistribMulActio
   ext rfl <| by simp [snd_pow_eq_sum]
 #align triv_sq_zero_ext.inl_pow TrivSqZeroExt.inl_pow
 
+example [Monoid R] [AddMonoid M] [DistribMulAction R M] [DistribMulAction Rᵐᵒᵖ M]
+    [SMulCommClass R Rᵐᵒᵖ M]
+    (a b : R) (m : M) : a •> m <• b = a •> (m <• b) := by
+  rw [← smul_comm]
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/TrivSqZeroExt.lean:665]: Also to be deleted, I assume

File: Mathlib/Analysis/PSeries.lean [hunk 5]
```diff
@@ -32,7 +32,7 @@ p-series, Cauchy condensation test
 /-!
 ### Cauchy condensation test
 
-In this section we prove the Cauchy condensation test: for `f : ℕ → ℝ≥0` or `f : ℕ → ℝ`,
+In this section we prove the Cauchy condensation test: for an antitone `f : ℕ → ℝ≥0` or `f : ℕ → ℝ`,
```
Comment (eric-wieser) [inline Mathlib/Analysis/PSeries.lean:35]: These comments are definitely out of scope for this PR!

File: Mathlib/GroupTheory/Perm/Cycle/Basic.lean [hunk 6]
```diff
@@ -735,15 +735,12 @@ theorem IsCycle.isConj (hσ : IsCycle σ) (hτ : IsCycle τ) (h : σ.support.car
...
-  erw [zpowersEquivZPowers_apply, zpowersEquivZPowers_apply]
-  dsimp
-    -- This used to be `rw`, but we need `erw` after leanprover/lean4#2644
-  erw [pow_succ, Perm.mul_apply]
+  erw [hσ.zpowersEquivSupport_symm_apply n]
+  simp only [← Perm.mul_apply, ← pow_succ']
+  erw [hσ.zpowersEquivSupport_symm_apply (n + 1)]
+  erw [zpowersEquivZPowers_apply, zpowersEquivZPowers_apply, zpowersEquivSupport_apply]
+  simp_rw [pow_succ', Perm.mul_apply]
+  rfl
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Perm/Cycle/Basic.lean:743]: Flagging that there seem to be rather a lot of `erw`s here; perhaps we should leave the comments referring to the lean4 change.

File: Mathlib/GroupTheory/Perm/Cycle/Type.lean [hunk 7]
```diff
@@ -487,8 +487,10 @@ theorem _root_.exists_prime_orderOf_dvd_card {G : Type*} [Group G] [Fintype G] (
   let σ :=
     Equiv.mk (f 1) (f (p - 1)) (fun s => by rw [hf2, add_tsub_cancel_of_le hp.out.one_lt.le, hf3])
       fun s => by rw [hf2, tsub_add_cancel_of_le hp.out.one_lt.le, hf3]
-  have hσ : ∀ k v, (σ ^ k) v = f k v := fun k v =>
-    Nat.rec (hf1 v).symm (fun k hk => Eq.trans (congr_arg σ hk) (hf2 k 1 v)) k
+  have hσ : ∀ k v, (σ ^ k) v = f k v := fun k =>
+    Nat.rec (fun v => (hf1 v).symm) (fun k hk v => by
+      rw [pow_succ, Perm.mul_apply, hk (σ v), Nat.succ_eq_one_add, ← hf2 1 k]
+      rfl) k
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Perm/Cycle/Type.lean:493]: A golfing opportunity for another reviewer (optional)

File: Mathlib/RingTheory/DedekindDomain/Ideal.lean [hunk 8]
```diff
@@ -1234,6 +1234,12 @@ theorem Ideal.IsPrime.mul_mem_pow (I : Ideal R) [hI : I.IsPrime] {a b : R} {n :
   exact Or.inr (Prime.pow_dvd_of_dvd_mul_right ((Ideal.prime_iff_isPrime hI0).mpr hI) _ ha h)
 #align ideal.is_prime.mul_mem_pow Ideal.IsPrime.mul_mem_pow
 
+theorem Ideal.IsPrime.mem_pow_mul (I : Ideal R) [hI : I.IsPrime] {a b : R} {n : ℕ}
+    (h : a * b ∈ I ^ n) : a ∈ I ^ n ∨ b ∈ I := by
+  rw [mul_comm] at h
+  rw [or_comm]
+  exact Ideal.IsPrime.mul_mem_pow _ h
+
```
Comment (eric-wieser) [inline Mathlib/RingTheory/DedekindDomain/Ideal.lean:1242]: Is this needed by this PR? It seems out of scope to me.

File: Mathlib/Algebra/Group/Commute/Defs.lean [hunk 9]
```diff
@@ -258,19 +258,20 @@ end Commute
 section Monoid
 variable {M : Type*} [Monoid M] {n : ℕ}
 
+-- Note (ACL) : this one has been changed to the initial meaning of pow_succ
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Commute/Defs.lean:261]: The `#align`s should be updated to reflect this. This comment should be removed; it's useful to a PR reviewer, but not to someone who finds the code later; so it belongs in the PR not the code!

File: Mathlib/Computability/RegularExpressions.lean [hunk 10]
```diff
@@ -146,8 +146,10 @@ theorem matches'_mul (P Q : RegularExpression α) : (P * Q).matches' = P.matches
 @[simp]
 theorem matches'_pow (P : RegularExpression α) : ∀ n : ℕ, (P ^ n).matches' = P.matches' ^ n
   | 0 => matches'_epsilon
-  | n + 1 => (matches'_mul _ _).trans <|
-      Eq.trans (congr_arg _ (matches'_pow P n)) (pow_succ _ _).symm
+  | n + 1 =>  (matches'_mul _ _).trans <| Eq.trans
```
Comment (eric-wieser) [inline Mathlib/Computability/RegularExpressions.lean:149]:
```suggestion
  | n + 1 => (matches'_mul _ _).trans <| Eq.trans
```

File: Mathlib/RingTheory/DedekindDomain/Ideal.lean (hunk 8 above)
Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/DedekindDomain/Ideal.lean:1242]: I needed it to make one proof work, it was too complicated to guess how the left/right, prime/unprime should be arranged while writing this lemma as the totally symmetric version of the one just above it was a trivial move.

Comment (eric-wieser) [reply Mathlib/RingTheory/DedekindDomain/Ideal.lean:1242]: Can you put this one in its own PR?

File: Mathlib/Algebra/Group/Commute/Defs.lean [hunk 11]
```diff
@@ -258,19 +258,20 @@ end Commute
...
 @[to_additive succ_nsmul']
-lemma pow_succ' (a : M) (n : ℕ) : a ^ (n + 1) = a ^ n * a :=
-  (pow_succ a n).trans (Commute.self_pow _ _)
-#align pow_succ' pow_succ'
-#align succ_nsmul' succ_nsmul'
+lemma pow_succ' (a : M) (n : ℕ) : a ^ (n + 1) = a * a ^ n := -- by
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Commute/Defs.lean:263]:
```suggestion
lemma pow_succ' (a : M) (n : ℕ) : a ^ (n + 1) = a * a ^ n :=
```

File: Mathlib/Algebra/Group/Commute/Defs.lean (hunk 9 above)
Comment (eric-wieser) [reply Mathlib/Algebra/Group/Commute/Defs.lean:261]:
```suggestion
```

File: Mathlib/Computability/RegularExpressions.lean [hunk 12]
```diff
@@ -146,8 +146,10 @@ theorem matches'_mul (P Q : RegularExpression α) : (P * Q).matches' = P.matches
 @[simp]
 theorem matches'_pow (P : RegularExpression α) : ∀ n : ℕ, (P ^ n).matches' = P.matches' ^ n
   | 0 => matches'_epsilon
-  | n + 1 => (matches'_mul _ _).trans <|
-      Eq.trans (congr_arg _ (matches'_pow P n)) (pow_succ _ _).symm
+  | n + 1 => (matches'_mul _ _).trans <| Eq.trans
+      (congrFun (congrArg HMul.hMul (matches'_pow P n)) (matches' P))
+      (pow_succ _ n).symm
+
```
Comment (eric-wieser) [inline Mathlib/Computability/RegularExpressions.lean:152]:
```suggestion
```

File: Mathlib/Computability/RegularExpressions.lean [hunk 13]
```diff
@@ -394,7 +396,8 @@ def map (f : α → β) : RegularExpression α → RegularExpression β
 protected theorem map_pow (f : α → β) (P : RegularExpression α) :
     ∀ n : ℕ, map f (P ^ n) = map f P ^ n
   | 0 => by dsimp; rfl
-  | n + 1 => (congr_arg (map f P * ·) (RegularExpression.map_pow f P n) : _)
+  | n + 1 => (congr_arg (· * map f P) (RegularExpression.map_pow f P n) : _)
+
```
Comment (eric-wieser) [inline Mathlib/Computability/RegularExpressions.lean:400]:
```suggestion
```

File: Mathlib/RingTheory/HahnSeries/Summable.lean [hunk 14]
```diff
@@ -103,7 +103,7 @@ theorem isPWO_iUnion_support_powers [LinearOrderedCancelAddCommMonoid Γ] [Ring
     rw [hn, SetLike.mem_coe]
     exact AddSubmonoid.zero_mem _
   · obtain ⟨i, hi, j, hj, rfl⟩ := support_mul_subset_add_support hn
-    exact SetLike.mem_coe.2 (AddSubmonoid.add_mem _ (AddSubmonoid.subset_closure hi) (ih hj))
+    exact SetLike.mem_coe.2 (AddSubmonoid.add_mem _  (ih hi) (AddSubmonoid.subset_closure hj))
```
Comment (eric-wieser) [inline Mathlib/RingTheory/HahnSeries/Summable.lean:106]:
```suggestion
    exact SetLike.mem_coe.2 (AddSubmonoid.add_mem _ (ih hi) (AddSubmonoid.subset_closure hj))
```

**Review by eric-wieser — COMMENTED** [review]
I think the main things needed to get this merged are:

* Fix the merge conflicts
* Run a benchmark

File: Mathlib/Algebra/Group/Defs.lean [hunk 15]
```diff
@@ -494,14 +494,14 @@ variable {M : Type u}
 Use instead `a ^ n`, which has better definitional behavior. -/
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:494]: There is a comment on line 491 that is presumably no longer relevant?

**Comment by AntoineChambert-Loir** [comment]
I am ready to handle the merge conflicts, if that doesn't bother your reviewing work.

**Comment by eric-wieser** [comment]
Please do!

**Comment by AntoineChambert-Loir** [comment]
(The first compilation won't work : the last update of mathlib introduced `add_sub_cancel_left` and `add_sub_cancel_right`; during the merge, without any filed compiled, I tried to guess which one I should insert, but I failed in some cases.)

**Comment by AntoineChambert-Loir** [comment]
!bench

**Comment by eric-wieser** [comment]
maintainer merge

I've gone though the whole diff, and any issues are pretty minor.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge
